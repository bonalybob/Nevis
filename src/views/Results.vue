<template>
  <base-layout>
    <div slot="menu">
      <router-link to="/dashboard" class="back">Back</router-link>
    </div>
    <div slot="main" class="main">
      <div
        v-for="course in courses"
        v-if="downloadsForCourse(course).length > 0"
        :key="course"
        class="card"
      >
        <h1>{{ course }}</h1>
        <table>
          <tbody>
            <tr>
              <th>Pos.</th>
              <th>Name</th>
              <th>Age Class</th>
              <th>Time</th>
            </tr>
          </tbody>
          <tbody is="transition-group" name="fade">
            <tr v-for="competitor of downloadsForCourse(course)" :key="competitor._id">
              <td>{{ competitor.position }}</td>
              <td>{{ competitor.name }}</td>
              <td>{{ competitor.ageClass }}</td>
              <td>{{ displayTime(competitor.result) }}</td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </base-layout>
</template>

<script>
import BaseLayout from '@/components/BaseLayout'

export default {
  components: {
    'base-layout': BaseLayout,
  },

  created: function () {
    if (this.$database.database === null) {
      this.$router.push('/')
      this.$messages.addMessage('Not Connected to the Database', 'error')
    }
  },

  methods: {
    downloadsForCourse: function (course) {
      if (this.downloads) {
        let downloads = this.downloads
        let currentPosition = 1
        downloads = downloads.filter(competitor => competitor.course === course)
        downloads = downloads.sort((a, b) => {
          if (typeof a.result === 'string' && typeof b.result === 'string') return a.result.length > b.result.length
          else if (typeof a.result === 'string') return true
          else if (typeof b.result === 'string') return false
          else return a.result > b.result
        })
        downloads.forEach(competitor => {
          if (competitor.nonCompetitive) competitor.position = 'n/c'
          else {
            competitor.position = currentPosition
            currentPosition += 1
          }
        })
        return downloads
      }
      else return []
    },

    displayTime: function (result) {
      if (typeof result !== 'number') return result
      else return this.$time.elapsed(result)
    },
  },

  asyncComputed: {
    courses: function () {
      return this.$database.getCourses()
        .then(data => data.map(course => course.doc.name))
        .catch(error => this.$messages.addMessage(error.message, 'error'))
    },

    downloads: function () {
      return this.$database.getDownloads()
        .catch(error => this.$messages.addMessage(error.message, 'error'))
    },
  },
}
</script>
