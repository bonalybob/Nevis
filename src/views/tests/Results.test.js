import { mount, shallowMount } from '@vue/test-utils'
import Results from '@/views/Results'

let MockDate = require('mockdate')

MockDate.set('2019-01-01')

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - No Results', () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Not Connected to the Database', () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: false,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
})

test('Get Courses - Success', async () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(['hello']),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.getCourses()
  expect(wrapper.vm.courses).toEqual(['hello'])
})

test('Get Courses - Error', async () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.getCourses()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Fetching Courses',
    'error'
  )
  expect(wrapper.vm.courses.length).toBe(0)
})

test('Get Results - Success', async () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(['hello']),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.getResults()
  expect(wrapper.vm.results).toEqual(['hello'])
})

test('Get Results - Error', async () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.getResults()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Fetching Results',
    'error'
  )
  expect(wrapper.vm.results.length).toBe(0)
})

test('Competitors On Course', () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  wrapper.setData({
    results: [
      { course: 1, errors: '', time: 35 },
      { course: 2, errors: '', time: 2 },
      { course: 1, errors: 'W2', time: 22 },
      { course: 2, errors: 'M4', time: 0 },
      { course: 1, errors: '', time: 15 },
      { course: 1, errors: '', time: 33, nonCompetitive: true },
    ],
  })
  expect(wrapper.vm.competitorsOnCourse(1)).toEqual([
    { course: 1, errors: '', time: 15, position: 1 },
    { course: 1, errors: '', time: 33, position: 'n/c', nonCompetitive: true },
    { course: 1, errors: '', time: 35, position: 2 },
    { course: 1, errors: 'W2', time: 22, position: 3 },
  ])
  expect(wrapper.vm.competitorsOnCourse(2)).toEqual([
    { course: 2, errors: '', time: 2, position: 1 },
    { course: 2, errors: 'M4', time: 0, position: 2 },
  ])
})

test('Get Event Details - Success', async () => {
  const wrapper = mount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(['hello']),
      },
      $route: { params: { id: 12 } },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })

  expect(await wrapper.vm.getEventDetails()).toBe('hello')
})

test('Get Event Details - Error', async () => {
  const wrapper = mount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 } },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.getEventDetails()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Fetching Event Data',
    'error'
  )
  expect(await wrapper.vm.getEventDetails()).toEqual(undefined)
})

test('Generate HTML', async () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  wrapper.setMethods({
    getEventDetails: jest.fn().mockResolvedValue({ name: 'a', date: 'b' }),
  })
  wrapper.setData({
    courses: [
      {
        id: 3,
        name: 'Short',
        length: 2530,
        climb: 35,
        type: 'linear',
        controls: '',
        event: 1,
      },
      {
        id: 5,
        name: 'Long',
        length: 7500,
        climb: 250,
        type: 'linear',
        controls: '101,102,103',
        event: 1,
      },
    ],
    results: [{ name: 'Bob', ageClass: 'M20', course: 3, errors: 'M12' }],
  })
  expect(await wrapper.vm.generateHTML()).toEqual(
    expect.stringContaining(`
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>a - Results</title>`)
  )

  expect(await wrapper.vm.generateHTML()).toEqual(
    expect.stringContaining(`<div class="course">
    <h2>Short</h2>
    <p>2.53km 35m</p>
    <table>
      <tr>
        <th>Pos.</th>
        <th>Name</th>
        <th class="ageClass">Age Class</th>
        <th class="club">Club</th>
        <th>Time</th>
      </tr>`)
  )
  expect(await wrapper.vm.generateHTML()).toEqual(
    expect.stringContaining(`<tr>
      <td></td>
      <td>Bob</td>
      <td class="ageClass">M20</td>
      <td class="club"></td>
      <td>M12</td>
    </tr>`)
  )

  expect(await wrapper.vm.generateHTML()).toEqual(
    expect.stringContaining(`</main>
    <footer>
      <p>Created at 2019-01-01 00:00:00 by Nevis </p>
    </footer>
  </body>
</html>`)
  )
})

test('Generate HTML - Error', async () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })

  wrapper.setMethods({
    getEventDetails: jest.fn().mockResolvedValue({ name: 'a', date: 'b' }),
    courseToHTML: jest.fn(() => {
      throw Error()
    }),
  })
  wrapper.setData({
    courses: [
      {
        id: 3,
        name: 'Short',
        length: 2530,
        climb: 35,
        type: 'linear',
        controls: '',
        event: 1,
      },
      {
        id: 5,
        name: 'Long',
        length: 7500,
        climb: 250,
        type: 'linear',
        controls: '101,102,103',
        event: 1,
      },
    ],
    results: [{ name: 'Bob', ageClass: 'M20', course: 3, errors: 'M12' }],
  })
  await wrapper.vm.generateHTML()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Generating HTML',
    'error'
  )
})

test('Save HTML - Cancelled', async () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $electron: {
        remote: {
          dialog: {
            showSaveDialog: jest.fn().mockResolvedValue({ canceled: true }),
          },
        },
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })

  await wrapper.vm.saveHTMLResults()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Saving Results',
    'error'
  )
})

test('Save HTML - Success', async () => {
  const wrapper = shallowMount(Results, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $electron: {
        remote: {
          dialog: {
            showSaveDialog: jest
              .fn()
              .mockResolvedValue({ canceled: false, filePath: 'test.html' }),
          },
        },
      },
      $fs: { writeFile: jest.fn() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })

  wrapper.setMethods({
    generateHTML: jest.fn().mockResolvedValue('a'),
  })

  await wrapper.vm.saveHTMLResults()
  expect(wrapper.vm.$fs.writeFile).toHaveBeenLastCalledWith('test.html', 'a', {
    encoding: 'utf8',
  })
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledWith(
    'Results Successfully Written'
  )
})
