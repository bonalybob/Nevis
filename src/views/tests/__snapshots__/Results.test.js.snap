// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - No Results 1`] = `
<main>
  <back-arrow-stub
    to="/events/12"
  />
   
  <h1
    class="mx-10 mb-1 px-1"
  >
    
    Results
  
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <button
      class="button"
    >
      Refresh
    </button>
     
    <button
      class="button"
    >
      Export HTML
    </button>
  </div>
   
</main>
`;
